AMO Swine wastewater treatment flowsheet

adam-a-a opened the pull request:
## Fixes/Resolves:
## Summary/Motivation:
Develop flowsheet for swine wastewater treatment along with some minimal sweep analysis to get things started.

## Changes proposed in this PR:
- revise unit models and associated tests as needed
-  add costing to unit model YAML files and costing package
- implement flowsheet and tests with minimal set of regression tests
- implement small subset of parameter sweeps in multi_sweep to initiate sensitivity analysis

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

adam-a-a committed:
starting flowsheet for 1503

adam-a-a committed:
slight addition to costing flowsheet

adam-a-a committed:
flowsheet solves; next step is to add unit and flowsheet costing

adam-a-a committed:
add some levelized costs and annual production rates at flowsheet level

adam-a-a committed:
finished adding costing to flowsheet; add cost methods for units

adam-a-a committed:
added cost method for AnMBR-MEC and unit cost parameters

adam-a-a committed:
test costing added for MBR-MEC

adam-a-a committed:
add pump electricity to vfa recovery and MBR-MEC; revise MBR-MEC test; add thermal energy req for vfa recovery yaml

adam-a-a committed:
added unit costing for vfa recovery unit with testing

adam-a-a committed:
add costing for other units except sedimentation tank; run flowsheet with costing for first check

adam-a-a committed:
Merge branch 'main' into 1503_flowsheet

adam-a-a committed:
tinker with display costing and see why costs so high- food waste flow too high causing high cost of cofermenter

adam-a-a committed:
checkpoint

adam-a-a committed:
add units to no_energy_electric_flow_vol_inlet list in test_unit_parameter_files

adam-a-a committed:
revise test_zero_order_costing to include heat in defined flows

adam-a-a committed:
start tests for 1503 flowsheet

adam-a-a committed:
add sedimentation subtype costing, test, and new cost metric for LCOT of influent flow

adam-a-a committed:
adding minimal testing for 1503 flowsheet

adam-a-a committed:
revise cofermentation unit, vfa unit, add cost blocks, change some flowsheet values

adam-a-a committed:
revise costing

adam-a-a committed:
fix test

adam-a-a committed:
add multi_sweep testing

adam-a-a committed:
add revenue for tipping fees associated with food waste taken as additional feed to cofermentation unit

adam-a-a committed:
update import on test

adam-a-a commented on watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/swine_wwt/swine_wwt.py:
```diff
@@ -0,0 +1,711 @@
+                to_units=m.fs.costing.base_currency / m.fs.costing.base_period,
+            )
+        ),
+        doc="Annual sludge disposal cost",
```
```suggestion
        doc="Annual food waste tipping fee revenue",
```

adam-a-a committed:
Update watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/swine_wwt/swine_wwt.py

adam-a-a committed:
Merge branch 'main' into 1503_flowsheet

andrewlee94 commented on watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/swine_wwt/swine_wwt.py:
```diff
@@ -0,0 +1,711 @@
+    initialize_system(m)
+
+    results = solve(m)
+    assert_optimal_termination(results)
```
Should we be adding a `try/except` on this (or perhaps doing a `check_optimal_termination` instead of `assert_optimal_termination`)? In the current form, if the result in not optimal the user will see an `AssertionError` with  default message which might not be that useful for explaining what happened.

adam-a-a replied:
We _could_, but then again, we've been doing this for all the AMO/zero-order model flowsheets, so we'd have to change this in all those flowsheets as well.  And currently, we're able to solve without issues, so if this were to fail, it would be because the user changed something in terms of fixing variables under `set_operating_conditions`. I think this serves its purpose as is for now.

Moreover, at the moment, our main objective is to get initial flowsheets implemented, produce techno-economic results, and iterate with teams on those results and model setup to determine next steps. @dangunter is working on a UI, so maybe this is something we'd want to consider at that level (i.e., a useful message for the user if/when result is not optimal).

adam-a-a replied:
@andrewlee94 I think replacing with `check_optimal_termination` in all AMO flowsheets in a separate PR would be good. Any other issues that need to be addressed now before you approve?

andrewlee94 commented on watertap/unit_models/zero_order/ion_exchange_zo.py:
```diff
@@ -41,9 +41,10 @@ def build(self):
-        self.eta_pump.set_value(0.8)
-        # mutable parameter; default value of 2 bar converted to feet head
-        self.lift_height.set_value(69.91052 * pyunits.feet)
+        if self.config.process_subtype != "clinoptilolite":
```
A minor point, but for future maintenance, it might be better to call out the exceptions instead:

```
if self.config.process_subtype != "clinoptilolite":
    pass
else:
    self.eta_pump.set_value(0.8)
    # mutable parameter; default value of 2 bar converted to feet head
    self.lift_height.set_value(69.91052 * pyunits.feet)
```

adam-a-a replied:
We could do that, except the logic would be the other way around. Basically, the default WT3 version specified said parameter values, while for this "clinoptilolite" subtype, I want to ignore those specifications. So, for now, I only check whether subtype is or isn't "clinoptilolite". If it isn't, keep the old specifications from WT3; if it is "clinoptilolite", don't touch the default parameter values which are typically used across most other models with pump electricity.

andrewlee94 replied:
Sorry, that is what I meant.

andrewlee94 reviewed (commented):
Only a couple of minor comments/suggestions,

adam-a-a commented on watertap/unit_models/zero_order/ion_exchange_zo.py:
```diff
@@ -41,9 +41,10 @@ def build(self):
+        if self.config.process_subtype != "clinoptilolite":
+            self.eta_pump.set_value(0.8)
+            # mutable parameter; default value of 2 bar converted to feet head
+            self.lift_height.set_value(69.91052 * pyunits.feet)
```
```suggestion
        if self.config.process_subtype == "clinoptilolite":
            pass
        else:
            self.eta_pump.set_value(0.8)
            # mutable parameter; default value of 2 bar converted to feet head
            self.lift_height.set_value(69.91052 * pyunits.feet)
```

adam-a-a committed:
Update watertap/unit_models/zero_order/ion_exchange_zo.py

adjust if-else for ion exchange

bknueven commented on watertap/core/zero_order_costing.py:
```diff
@@ -856,6 +927,113 @@ def cost_coag_and_floc(blk):
+        parameter_dict = blk.unit_model.config.database.get_unit_operation_parameters(
+            blk.unit_model._tech_type, subtype=blk.unit_model.config.process_subtype
+        )
+        parameter_dict
```
```suggestion 

```

bknueven commented on watertap/data/techno_economic/tests/test_unit_parameter_files.py:
```diff
@@ -102,6 +103,8 @@ def test_unit_parameter_files(tech):
         "well_field",
         "mbr",
+        "vfa_recovery",
+        "cofermentation",
```
It would be preferable to maintain the alphabetical order here, so it's easy to know if something is in the list by inspection.

bknueven commented on watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/swine_wwt/multi_sweep.py:
```diff
@@ -0,0 +1,100 @@
+
+
+if __name__ == "__main__":
+    results, sweep_params, m = run_analysis(*sys.argv[1:])
```
It would be nice to put a basic help message here, e.g.,
```suggestion
    if len(sys.argv) == 1:
        print("Usage: python multi_sweep.py case_number number_of_samples interpolate_nan_outputs")
        print(f"Results will be written to {os.path.dirname(os.path.abspath(__file__))}")
    else:
        results, sweep_params, m = run_analysis(*sys.argv[1:])
```

bknueven reviewed (commented):
I have a few minor suggestions, mostly relating to maintenance.

At a high level, the zero order costing package has too much repeated code, and this PR is not helping matters. But it should be cleaned up subsequently.

adam-a-a committed:
Update watertap/core/zero_order_costing.py

Co-authored-by: bknueven <30801372+bknueven@users.noreply.github.com>

adam-a-a committed:
make list alphabetical order

adam-a-a committed:
add message to main in multisweep

adam-a-a committed:
Merge branch 'main' into 1503_flowsheet

adam-a-a commented:
> I have a few minor suggestions, mostly relating to maintenance.
> 
> At a high level, the zero order costing package has too much repeated code, and this PR is not helping matters. But it should be cleaned up subsequently.

@bknueven I addressed your suggestions. As you mentioned, we could refactor the zero order costing package in a subsequent PR to eliminate repeated code.
